src/main/kotlin/it/pagopa/wallet/warmup/ControllersWarmup.kt

```diff
@@ -0,0 +1,65 @@
+package it.pagopa.wallet.warmup
+
+import it.pagopa.wallet.warmup.annotations.WarmupFunction
+import kotlin.reflect.full.declaredMemberFunctions
+import kotlin.reflect.full.hasAnnotation
+import kotlin.system.measureTimeMillis
+import org.slf4j.LoggerFactory
+import org.springframework.beans.factory.getBeansWithAnnotation
+import org.springframework.context.ApplicationListener
+import org.springframework.context.event.ContextRefreshedEvent
+import org.springframework.stereotype.Component
+import org.springframework.util.ClassUtils
+import org.springframework.web.bind.annotation.RestController
+
+@Component
+class ControllersWarmup : ApplicationListener<ContextRefreshedEvent> {
+
+    private val logger = LoggerFactory.getLogger(this.javaClass)
+    override fun onApplicationEvent(event: ContextRefreshedEvent) {
```

Comment:
Can ContextRefreshedEvent be triggered more than once during the app lifecycle? If yes, should we execute warmup functions only once or every time?

Reply:
This event is raised automatically by Spring once application context is starter and every time the `ConfigurableApplicationContext .refresh()` method is programmatically called against (that is not our case since we do not use ConfigurableApplicationContext explicitly.  [see here](https://www.baeldung.com/spring-context-events#1-contextrefreshedevent).

For actual implementation this event is fired once before making application ready to accept external traffic (actuator endpoints are not enabled, f.e. so application is not seen as ready before all warmup method have completed their executions).

In any case, warmup requests have been implemented to have no side-effects (no write to DB, just api calls to external services) so are idempotent and there would be no problem if executed more than once

Reply:
It's ok to me if we're sure this event is triggered only once with the current app implementation and the warmup functions don't have side effects